doc/manpages/stl2gts.1
.TH stl2gts 1 "June 2, 2008" "" "GNU Triangulated Surface utils"

.SH NAME
stl2gts \- convert an STL file to GTS format.

.SH SYNOPSIS
.B stl2gts
.RI [\| OPTIONS \|]\ <\  input.stl\  >\  output.gts

.SH DESCRIPTION
This manual page documents briefly the
.B stl2gts
command.

.SH OPTIONS
These programs follow the usual GNU command line syntax, with long
options starting with two dashes (`-').
A summary of options is included below.
.TP
.BR \-r ,\  \-\-revert
Revert face normals.
.TP
.BR \-n ,\  \-\-nomerge
Do not merge vertices.
.TP
.BR \-v ,\  \-\-verbose
Display surface statistics.
.TP
.BR \-h ,\  \-\-help
Display the help and exit.

.SH AUTHOR
stl2gts was written by Stephane Popinet <popinet@users.sourceforge.net>.
.PP
This manual page was written by Ruben Molina <rmolina@udea.edu.co>,
for the Debian project (but may be used by others).


NEWS
29/03/2006:

Release of version 0.7.6

Important changes:

- Debian snapshot package creation.
- LC_ALL is set to POSIX in all executables to avoid problems with
locale-dependent floating-point conventions.

17/10/2005:

Release of version 0.7.4

Important changes:

- Switched to darcs for version control (the CVS tree is obsolete).

Bug Fixes:

- For constrained Delaunay, closest face for glib >= 2.4 did not use planar distance.
- Several fixes for gts_delaunay_add_constraint.
- Fix for degenerate case in isotetra_bcl().
- Fix for comments not being fully ignored by GtsFile.

Minor changes and features:

- delaunay and happrox are installed.
- Fast triangle-box intersection test by Tomas Akenine-Moller.
- New function gts_triangle_interpolate_height().
- binary STL files work on win32.

18/10/2004:

Release of version 0.7.3

New features/important changes include:

- Autodetection and support for glib >= 2.4 (but glib >= 1.2.8 is
  still supported): glib >= 2.0 and < 2.4 is not supported (send me
  patches!).
- pkgconfig preliminary support
- Mingwin support.
- Changes to the interface of gts_graph_read().

Bug Fixes:

- Constrained Delaunay Triangulation guess selection
- Boolean operations: Now uses SOS for segment/triangle intersections.

15/05/2004:

Release of version 0.7.2

New features includes:

- gts2dxf and gts2stl filters.
- New GtsColorVertex class.
- "Dual" isosurface calculation by Tim F.
- New "happrox" algorithm to approximate terrain models.
- New GtsMatrix implementation (quaternions) by Wayne Gramlich and
  Michael Loehr.
- Attributes inheritance when computing intersections.

Bug fixes:

- Major changes to inner loops triangulation for boolean operations.
- compilation on SGI Altix
- Fixed gts.def generation for Windows support.
- Cygwin support.

08/01/2003:

Release of version 0.7.1

New features includes:

- Sphere tessellation by Jerome Benoit.

As well as a number of bug and compilation fixes.

The interface to the gts_surface_refine() function has been made more
generic.

08/11/2002:

Release of version 0.7.0

Boolean operations have been almost entirely rewritten and now use a
Simulation of Simplicity (SoS) technique to deal with degenerate
cases. This should fix most of the problems encountered when dealing
with "simple" geometric objects (i.e. cube/cube intersections with
coplanar faces etc...) Almost all the test cases in the test suite
(including significant tests submitted by users) now pass. Difficult
test cases like an object intersecting with a simplified version of
itself (created using coarsen for example) work fine despite the huge
number of degenerate cases (coincident and nearly coincident edges,
faces, vertices).

There is still a caveat however for intersections which creates
multiple closed loops contained within one another (note that this
wasn't a problem in the previous version). This needs to be fixed at
some point.

Another problem is that the surfaces resulting from a boolean
operation while guaranteed to be topologically consistent may contain
degenerate edges/faces which will prevent them from being used in a
subsequent boolean operation. The degenerate entities need first to be
eliminated (through edge/face collapse).

The gts_point_is_inside_surface() function has been rewritten using SoS.

A new stl2gts tool provides conversion between STL (stereolithography)
format and GTS format.

Ray Jones has added new functions to compute gaussian curvature,
principal curvature directions etc...

The gts_file_close() function call has been replaced by a
gts_file_destroy() function which does not close the file. This is
better left to the caller. Your application might need a slight
modification there.

The whole automake/autoconf stuff has been upgraded to use version
1.6.3 of automake.

Gert Wollny has added a marching tetrahedra isosurface construction
technique which should avoid topological defects in isosurfaces
(created by inconsistencies inherent to the marching cube technique).

A number of bugs have been fixed (thanks to bug reporters).

08/10/2001:

Release of version 0.5.0.

Configuration and compilation have been cleaned and updated. Now
builds both shared and static libraries.

A new `gts-config' script allows simpler use of the library (a la
glib), together with a gts.m4 macro definition for your own autoconf
usage.

A number of new features have been added:

- Configurable out-of-core simplification framework for surfaces
larger than the main memory.

- Conforming constrained Delaunay triangulation and Delaunay
refinement by addition of Steiner points.

- Simple implementation of FIFOs (First-In First-out queues).

- New graph objects and operations on graphs. In particular efficient
graph partitioning routines.

- "Stabbing" functions have been added to bounding box trees, which
now allows for efficient testing whether a point lies inside or outside
a surface.

A number of bug fixes and improvements have also been added:

- More generic object methods for surfaces.

- Easily definable inheritance of face attributes (in particular for
boolean operations).

- Bug fixes for boolean operations.

The srf2oogl converter has been renamed to gts2oogl and an 'isolines'
functionality was added.

31/01/2001: 

Release of version 0.4.0. Much has happened since 0.3.0. A
new object structure inspired by GTK allows to use classical
inheritance mechanisms and virtual functions. As a consequence the
object hierarchy has been restructured and now uses points, vertices,
segments, edges, triangles, faces and surfaces.

The Delaunay triangulation has been entirely re-written and is now
fully-dynamic (insertion or deletion of vertices). The constrained
Delaunay triangulation is semi-dynamic (insertion of vertices and
constraints).

Boolean operations between surfaces have also been re-written and are
more robust (and also benefit from a much more robust implementation
of the Delaunay triangulation).


tools/stl2gts.c
/* GTS - Library for the manipulation of triangulated surfaces
 * Copyright (C) 1999 Stéphane Popinet
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Library General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.	 See the GNU
 * Library General Public License for more details.
 *
 * You should have received a copy of the GNU Library General Public
 * License along with this library; if not, write to the
 * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
 * Boston, MA 02111-1307, USA.
 */

#include <math.h>
#include <stdlib.h>
#include <locale.h>
#include <string.h>
#include "config.h"
#ifdef HAVE_GETOPT_H
#  include <getopt.h>
#endif /* HAVE_GETOPT_H */
#ifdef HAVE_UNISTD_H
#  include <unistd.h>
#endif /* HAVE_UNISTD_H */
#include "gts.h"

#ifdef NATIVE_WIN32
#  include <fcntl.h>
#  include <io.h>
#endif 

static GPtrArray * stl_read (FILE * fp)
{
  GPtrArray * a = g_ptr_array_new ();
  char tag[6];

  fgets (tag, 6, fp);
  rewind (fp);
  if (!strcmp (tag, "solid")) { /* ASCII file */
    GtsFile * f = gts_file_new (fp);

    while (f->type == GTS_STRING) {
      if (!strcmp (f->token->str, "vertex")) {
	gdouble x, y, z;

	gts_file_next_token (f);
	if (f->type != GTS_INT && f->type != GTS_FLOAT) {
	  fprintf (stderr, 
		   "Input file is not a valid STL file\n"
		   "stdin:%d:%d: expecting a number (x-coordinate)\n",
		   f->line, f->pos);
	  exit (1);
	}
	x = strtod (f->token->str, NULL);

	gts_file_next_token (f);
	if (f->type != GTS_INT && f->type != GTS_FLOAT) {
	  fprintf (stderr, 
		   "Input file is not a valid STL file\n"
		   "stdin:%d:%d: expecting a number (y-coordinate)\n",
		   f->line, f->pos);
	  exit (1);
	}
	y = strtod (f->token->str, NULL);

	gts_file_next_token (f);
	if (f->type != GTS_INT && f->type != GTS_FLOAT) {
	  fprintf (stderr, 
		   "Input file is not a valid STL file\n"
		   "stdin:%d:%d: expecting a number (z-coordinate)\n",
		   f->line, f->pos);
	  exit (1);
	}
	z = strtod (f->token->str, NULL);

	g_ptr_array_add (a, gts_vertex_new (gts_vertex_class (), x, y, z));
      }
      else if (!strcmp (f->token->str, "endsolid"))
	break;
      gts_file_first_token_after (f, '\n');
    }
    gts_file_destroy (f);
  }
  else { /* binary file */
    guint32 nf;
    gchar header[80];
    guint i;

#ifdef NATIVE_WIN32
    _setmode ( _fileno (fp), _O_BINARY);
#endif 

    if (fread (header, sizeof (gchar), 80, fp) != 80) {
      fprintf (stderr, "Input file is not a valid STL file\n"
	       "stdin: incomplete header\n");
      exit (1);
    }
    if (fread (&nf, sizeof (guint32), 1, fp) != 1) {
      fprintf (stderr, "Input file is not a valid STL file\n"
	       "stdin: missing number of facets\n");
      exit (1);
    }
    i = nf;
    while (i > 0) {
      gfloat x, y, z;
      guint j;
      guint16 attbytecount;

      if (fread (&x, sizeof (gfloat), 1, fp) != 1) {
	fprintf (stderr, "Input file is not a valid STL file\n"
	       "stdin: missing normal x-coordinate\n");
	exit (1);
      }
      if (fread (&y, sizeof (gfloat), 1, fp) != 1) {
	fprintf (stderr, "Input file is not a valid STL file\n"
	       "stdin: missing normal y-coordinate\n");
	exit (1);
      }
      if (fread (&z, sizeof (gfloat), 1, fp) != 1) {
	fprintf (stderr, "Input file is not a valid STL file\n"
	       "stdin: missing normal z-coordinate\n");
	exit (1);
      }

      for (j = 0; j < 3; j++) {
	if (fread (&x, sizeof (gfloat), 1, fp) != 1) {
	  fprintf (stderr, "Input file is not a valid STL file\n"
		   "stdin: missing vertex x-coordinate\n");
	  exit (1);
	}
	if (fread (&y, sizeof (gfloat), 1, fp) != 1) {
	  fprintf (stderr, "Input file is not a valid STL file\n"
		   "stdin: missing vertex y-coordinate\n");
	  exit (1);
	}
	if (fread (&z, sizeof (gfloat), 1, fp) != 1) {
	  fprintf (stderr, "Input file is not a valid STL file\n"
		   "stdin: missing vertex z-coordinate\n");
	  exit (1);
	}
	g_ptr_array_add (a, gts_vertex_new (gts_vertex_class (), x, y, z));
      }

      if (fread (&attbytecount, sizeof (guint16), 1, fp) != 1) {
	fprintf (stderr, "Input file is not a valid STL file\n"
	       "stdin: missing attribute byte count\n");
	exit (1);
      }
      i--;
    }
  }

  return a;
}

static void vertices_merge (GPtrArray * stl, gdouble epsilon)
{
  GPtrArray * array;
  GNode * kdtree;
  guint i;

  array = g_ptr_array_new ();
  for (i = 0; i < stl->len; i++)
    g_ptr_array_add (array, stl->pdata[i]);
  kdtree = gts_kdtree_new (array, NULL);
  g_ptr_array_free (array, TRUE);

  for (i = 0; i < stl->len; i++) {
    GtsVertex * v = stl->pdata[i];

    if (!GTS_OBJECT (v)->reserved) { /* Do something only if v is active */
      GtsBBox * bbox;
      GSList * selected, * j;

      /* build bounding box */
      bbox = gts_bbox_new (gts_bbox_class (),
			   v, 
			   GTS_POINT (v)->x - epsilon,
			   GTS_POINT (v)->y - epsilon,
			   GTS_POINT (v)->z - epsilon,
			   GTS_POINT (v)->x + epsilon,
			   GTS_POINT (v)->y + epsilon,
			   GTS_POINT (v)->z + epsilon);

      /* select vertices which are inside bbox using kdtree */
      j = selected = gts_kdtree_range (kdtree, bbox, NULL);
      while (j) {
	GtsVertex * sv = j->data;

	if (sv != v && !GTS_OBJECT (sv)->reserved)
	  GTS_OBJECT (sv)->reserved = v; /* mark sv as inactive */
	j = j->next;
      }
      g_slist_free (selected);
      gts_object_destroy (GTS_OBJECT (bbox));
    }
  }

  gts_kdtree_destroy (kdtree);

  /* destroy inactive vertices */

  /* we want to control vertex destruction */
  gts_allow_floating_vertices = TRUE;

  for (i = 0; i < stl->len; i++) {
    GtsVertex * v = stl->pdata[i];

    if (GTS_OBJECT (v)->reserved) { /* v is inactive */
      stl->pdata[i] = GTS_OBJECT (v)->reserved;
      gts_object_destroy (GTS_OBJECT (v));
    }
  }

  gts_allow_floating_vertices = FALSE; 
}

static void add_stl (GtsSurface * s, GPtrArray * stl)
{
  guint i;

  for (i = 0; i < stl->len/3; i++) {
    GtsEdge * e1 = GTS_EDGE (gts_vertices_are_connected (stl->pdata[3*i],
							 stl->pdata[3*i + 1]));
    GtsEdge * e2 = GTS_EDGE (gts_vertices_are_connected (stl->pdata[3*i + 1],
							 stl->pdata[3*i + 2]));
    GtsEdge * e3 = GTS_EDGE (gts_vertices_are_connected (stl->pdata[3*i + 2],
							 stl->pdata[3*i]));

    if (e1 == NULL && stl->pdata[3*i] != stl->pdata[3*i + 1])
      e1 = gts_edge_new (s->edge_class, 
			 stl->pdata[3*i], stl->pdata[3*i + 1]);
    if (e2 == NULL && stl->pdata[3*i + 1] != stl->pdata[3*i + 2])
      e2 = gts_edge_new (s->edge_class, 
			 stl->pdata[3*i + 1], stl->pdata[3*i + 2]);
    if (e3 == NULL && stl->pdata[3*i + 2] != stl->pdata[3*i])
      e3 = gts_edge_new (s->edge_class, 
			 stl->pdata[3*i + 2], stl->pdata[3*i]);
    if (e1 != NULL && e2 != NULL && e3 != NULL)
      gts_surface_add_face (s, gts_face_new (s->face_class, e1, e2, e3));
    else
      fprintf (stderr, "stl2gts: warning: ignoring degenerate facet!");
  }
}

int main (int argc, char * argv[])
{
  int c = 0;
  gboolean verbose = FALSE;
  gboolean nomerge = FALSE;
  gboolean revert  = FALSE;
  GPtrArray * stl;
  GtsSurface * s;

  if (!setlocale (LC_ALL, "POSIX"))
    g_warning ("cannot set locale to POSIX");

  while (c != EOF) {
#ifdef HAVE_GETOPT_LONG
    static struct option long_options[] = {
      {"revert", no_argument, NULL, 'r'},
      {"nomerge", no_argument, NULL, 'n'},
      {"help", no_argument, NULL, 'h'},
      {"verbose", no_argument, NULL, 'v'},
      { NULL }
    };
    int option_index = 0;
    switch ((c = getopt_long (argc, argv, "hvnr",
			      long_options, &option_index))) {
#else /* not HAVE_GETOPT_LONG */
    switch ((c = getopt (argc, argv, "hvnr"))) {
#endif /* not HAVE_GETOPT_LONG */
    case 'r': /* revert */
      revert = TRUE;
      break;
    case 'n': /* nomerge */
      nomerge = TRUE;
      break;
    case 'h': /* help */
      fprintf (stderr,
             "Usage: stl2gts [OPTION]... < input.stl > output.gts\n"
	     "Convert an STL file to GTS format.\n"
	     "\n"
	     "  -r,     --revert       revert face normals\n"
	     "  -n,     --nomerge      do not merge vertices\n"
	     "  -v,     --verbose      display surface statistics\n"
	     "  -h,     --help         display this help and exit\n"
	     "\n"
	     "Report bugs to %s\n",
	     GTS_MAINTAINER);
      return 0;
      break;
    case 'v':
      verbose = TRUE;
      break;
    case '?': /* wrong options */
      fprintf (stderr, "Try `stl2gts --help' for more information.\n");
      return 1;
    }
  }

  stl = stl_read (stdin);
  if (!nomerge)
    vertices_merge (stl, 0.);
  s = gts_surface_new (gts_surface_class (),
		       gts_face_class (),
		       gts_edge_class (),
		       gts_vertex_class ());
  add_stl (s, stl);
  if (revert)
    gts_surface_foreach_face (s, (GtsFunc) gts_triangle_revert, NULL);
  if (verbose)
    gts_surface_print_stats (s, stderr);
  gts_surface_write (s, stdout);

  return 0;
}
